File: 2210-count-hills-and-valleys-in-an-array/2210-count-hills-and-valleys-in-an-array.java
class Solution {
    public int countHillValley(int[] nums) {
        int n = nums.length, ans = 0, start = 1;

        // skip all initial equal numbers
        while(start < n && nums[start] == nums[start-1]) start++;
        if(start == n) return 0;

        // find all maxima and minima
        var dir = nums[start] > nums[start-1] ? 1 : -1;    // 1 = up, -1 = down
        for(var i=start+1; i<n; i++){
            if(nums[i] == nums[i-1]) continue;  // equal neighbour
            
            var newDir = nums[i] > nums[i-1] ? 1 : -1;
            if(newDir != dir) ans++;    // found maxima or minima
            dir = newDir;
        }

        return ans;
    }
}

File: 2210-count-hills-and-valleys-in-an-array/2210-count-hills-and-valleys-in-an-array.py
class Solution:
    def countHillValley(self, nums: List[int]) -> int:
        n, ans, start = len(nums), 0, 1

        # skip all initial equal numbers
        while start < n and nums[start] == nums[start-1]: start += 1
        if start == n: return 0

        # find all maxima and minima
        dir = 1 if nums[start] > nums[start-1] else -1    # 1 = up, -1 = down
        for i in range(start+1, n):
            if nums[i] == nums[i-1]: continue  # equal neighbour
            
            newDir = 1 if nums[i] > nums[i-1] else -1
            if newDir != dir: ans += 1    # found maxima or minima
            dir = newDir

        return ans